src/threshold_sign.rs

```diff
@@ -99,26 +101,42 @@ impl<N: NodeIdT> DistAlgorithm for ThresholdSign<N> {
 
 impl<N: NodeIdT> ThresholdSign<N> {
     /// Creates a new instance of `ThresholdSign`, with the goal to collaboratively sign `msg`.
-    pub fn new<M: AsRef<[u8]>>(netinfo: Arc<NetworkInfo<N>>, msg: M) -> Self {
+    pub fn new(netinfo: Arc<NetworkInfo<N>>) -> Self {
         ThresholdSign {
             netinfo,
-            msg_hash: hash_g2(msg),
+            msg_hash: None,
             received_shares: BTreeMap::new(),
             had_input: false,
             terminated: false,
         }
     }
 
+    /// Sets msg_hash and resets internal state accordingly.
+    pub fn set_message<M: AsRef<[u8]>>(&mut self, msg: M) -> Result<()> {
```

Comment:
Now that I think of it, "message" was a bad choice, since it's easy to confuse with the network messages being handled in `handle_message`. My bad! :grimacing: 
Maybe `set_document`?